### ISSUE TITLE
NULL pointer access behaviour

### ISSUE THREAD

**Issue by qhuo** [mit-pdos/xv6-riscv#272]
It seems that NULL pointer is mapped into user space pagetable, such that the below code works!

```
// Test NULL pointer access.
// FIXME: this shouldn't pass.
void
nullptr(char *s)
{
  const char* const pz = (const char*) 0;

  const char c = *pz;

  printf("*(const char*)0 == %x\n", (int) c);
}
```

Would it be better to prevent this? Ideally we shall map a guard map at virtual address 0.

I can make a change if there are no objections.

**Comment by jrmenzel** [comment]
Yes, you would have to map the apps one page higher. Right now page 0 is read only, so writing to a NULL pointer will cause an exception. To also case an exception on read you should have nothing mapped there (an empty mapped page with no access rights works too but is wasted memory).

Note that a few places in the code assume that the apps starts at 0, the linker script, the processes size, how a processes memory gets copied in fork etc.

This fix would also make a nice programming assignment, so it could be intentional to have apps at page 0.

**Comment by qhuo** [comment]
Hi, thanks for your reply.

Yes, I agree it would be a good exercise. I would be interested to implement this.

**Comment by kaashoek** [comment]
The page table lab (https://pdos.csail.mit.edu/6.1810/2024/labs/pgtbl.html) has this as an optional challenge exercise.
